docs/Training-Plugins.md

```diff
@@ -0,0 +1,55 @@
+# Customizing Training via Plugins
+
+ML-Agents provides support for running your own python implementations of specific interfaces during the training
+process. These interfaces are currently fairly limited, but will be expanded in the future.
+
+** Note ** Plugin interfaces should currently be considered "in beta", and they may change in future releases.
+
+## How to Write Your Own Plugin
+[This video](https://www.youtube.com/watch?v=fY3Y_xPKWNA) explains the basics of how to create a plugin system using
+setuptools, and is the same approach that ML-Agents' plugin system is based on.
+
+The `ml-agents-plugin-examples` directory contains a reference implementation of each plugin interface, so it's a good
+starting point.
+
+### setup.py
+If you don't already have a `setup.py` file for your python code, you'll need to add one. `ml-agents-plugin-examples`
+has a [minimal example](../ml-agents-plugin-examples/setup.py) of this.
+
+In the call to `setup()`, you'll need to add to the `entry_points` dictionary for each plugin interface that you
+implement. The form of this is `{entry point name}={plugin module}:{plugin function}`. For example, in
+ `ml-agents-plugin-examples`:
+```python
+"mlagents.stats_writer": [
+    "example=mlagents_plugin_examples.example_stats_writer:get_example_stats_writer"
+]
+```
+* `mlagents.stats_writer` is the name of the plugin interface. This must be one of the provided interfaces (see below).
+* `example` is the plugin implementation name. This can be anything.
+* `mlagents_plugin_examples.example_stats_writer` is the plugin module. This points to the module where the
+plugin registration function is defined.
+* `get_example_stats_writer` is the plugin registration function. This is called when running `mlagents-learn`. The
+arguments and expected return type for this are different for each plugin interface.
+
+### Local Installation
+Once you've defined `entry_points` in your `setup.py`, you will need to run
+```
+pip install -e [path to your plugin code]
+```
+in the same python virtual environment that you have `mlagents` installed.
+
+## Plugin Interfaces
+
+### StatsWriter
```

Comment:
It is not obvious that `mlagents.stats_writer` is a magic string. Can this be made more explicit here ?

Reply:
I moved mlagents.stats_writer to a constant in `mlagents.plugins.__init__.py`